src/components/ConnectWallet/ConnectWalletDialog.tsx

```diff
@@ -45,19 +57,90 @@ const StyledNoWalletText = styled(ThemedText.Subhead1)`
 interface ButtonProps {
   walletName?: string
   logoSrc?: string
-  caption?: string
+  connection?: Web3Connection
   onClick: () => void
 }
 
-function WalletConnectButton({ walletName, logoSrc, caption, onClick }: ButtonProps) {
+const wcQRUriAtom = atom<string | undefined>(undefined)
+
+function WalletConnectButton({ walletName, logoSrc, connection: wcTileConnection, onClick }: ButtonProps) {
+  const [walletConnect] = wcTileConnection as [WalletConnect, Web3ReactHooks]
+  const [error, setError] = useState(undefined)
+
+  const [QRUri, setQRUri] = useAtom(wcQRUriAtom)
+  const [qrCodeSvg, setQrCodeSvg] = useState<string>('')
+
+  useEffect(() => {
+    if (QRUri) {
+      formatQrCodeImage(QRUri)
+    } else {
+      walletConnect.activate().catch(setError)
+    }
+  }, [QRUri, walletConnect])
+
+  useEffect(() => {
+    // Log web3 errors
+    if (error) {
+      console.error('web3 error:', error)
+    }
+  }, [error])
+
+  useEffect(() => {
+    const disconnectListener = async (err: Error | null, _: any) => {
+      if (err) console.warn(err)
+      // Clear saved QR URI after disconnection
+      setQRUri(undefined)
+      walletConnect.deactivate()
+    }
+    walletConnect.provider?.connector.on('disconnect', disconnectListener)
+
+    // Need both URI event listeners
+    walletConnect.events.on(URI_AVAILABLE, async (uri: string) => {
+      if (uri) {
+        setQRUri(uri)
+        await formatQrCodeImage(uri)
+      }
+    })
+
+    const uriListener = async (err: Error | null, payload: any) => {
+      if (err) console.warn(err)
+      const uri: string = payload.params[0]
+      if (uri) {
+        setQRUri(uri)
+        await formatQrCodeImage(uri)
+      }
+    }
+    walletConnect.provider?.connector.on('display_uri', uriListener)
+
+    return () => {
+      walletConnect.events.off(URI_AVAILABLE)
+      ;(walletConnect.provider?.connector as unknown as EventEmitter | undefined)?.off('display_uri', uriListener)
+    }
+  })
+
+  async function formatQrCodeImage(uri: string) {
+    let result = ''
+    const dataString = await QRCode.toString(uri, { margin: 0, type: 'svg' })
+    if (typeof dataString === 'string') {
+      result = dataString.replace('<svg', `<svg class="walletconnect-qrcode_tile" alt="WalletConnect" width="100"`)
```

Comment:
The viewbox is set to `0 0 53 53`, so the width must be a multiple of 53 (ie 106) to prevent smushing the image (sorry idk the technical word). Does `QRCode` have an option to use `50`?